### PR TITLE
declared variable pattern support for `unnecessary_nullable_for_final_variable_declarations`

### LINKED ISSUE #4181: `unnecessary_nullable_for_final_variable_declarations` should have tests with patterns
Child issue of https://github.com/dart-lang/linter/issues/3624

I think there should be a test case like:

```dart
final (List<int>? a, List<int?> b, num? c) = ([], [], 1);
```

in which the lint rule would report that all three of the types are unnecessarily nullable.

### PR DESCRIPTION
**Pull request by pq**
Fixes #4181

/cc @bwilkerson

### CODE REVIEW

File: lib/src/rules/unnecessary_nullable_for_final_variable_declarations.dart [hunk 1]
```diff
@@ -63,6 +65,19 @@ class _Visitor extends SimpleAstVisitor<void> {
   final LinterContext context;
   _Visitor(this.rule, this.context);
 
+  void check(AstNode? element) {
```
Comment (srawlins) [inline lib/src/rules/unnecessary_nullable_for_final_variable_declarations.dart:68]: I think this is only called with non-nullable `DartPattern`s.

File: test/rules/unnecessary_nullable_for_final_variable_declarations_test.dart [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
```
Comment (srawlins) [inline test/rules/unnecessary_nullable_for_final_variable_declarations_test.dart:1]: 2023 😄

**Review by srawlins — APPROVED** [review]
❤️

File: lib/src/rules/unnecessary_nullable_for_final_variable_declarations.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/unnecessary_nullable_for_final_variable_declarations.dart:68]: Good catch!

**Pull request merged**